governance-chat/program/src/processor.rs

```diff
@@ -0,0 +1,114 @@
+//! Program processor
+
+use crate::{
+    error::GovernanceChatError,
+    instruction::GovernanceChatInstruction,
+    state::{assert_is_valid_chat_message, ChatMessage, GovernanceChatAccountType, MessageBody},
+    tools::account::create_and_serialize_account,
+};
+use borsh::BorshDeserialize;
+
+use solana_program::{
+    account_info::{next_account_info, AccountInfo},
+    clock::Clock,
+    entrypoint::ProgramResult,
+    msg,
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    sysvar::Sysvar,
+};
+use spl_governance::state::{
+    governance::get_governance_data, proposal::get_proposal_data_for_governance,
+    token_owner_record::get_token_owner_record_data_for_realm,
+};
+
+/// Processes an instruction
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    input: &[u8],
+) -> ProgramResult {
+    let instruction = GovernanceChatInstruction::try_from_slice(input)
+        .map_err(|_| ProgramError::InvalidInstructionData)?;
+
+    match instruction {
+        GovernanceChatInstruction::PostMessage { body } => {
+            msg!("GOVERNANCE-CHAT-INSTRUCTION: PostMessage");
+            process_post_message(program_id, accounts, body)
+        }
+    }
+}
+
+/// Processes PostMessage instruction
+pub fn process_post_message(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    body: MessageBody,
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let governance_program_info = next_account_info(account_info_iter)?; // 0
```

Comment:
I wonder if the associated governance program id should be hard-coded somewhere in the program.  Otherwise, people from governance A can post messages in the chat program for governance B, right? Unless this is meant to be the one chat program for all governances

Reply:
Yes you are right, in the current version the chat program is meant to be a singleton (one for all governance clones) because there is no security risk. 
However this might change in the future if we for example start using some reputation or community engagement score for governance rights and then each governance clone would need a corresponding chat clone. In that case I was thinking about putting the chat clone id on realm config which could then be validated against.